### ISSUE TITLE
Problems with generalization on let_mono branch (PR #33)

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/inferno#34]
I've been investigating why `scoped_tyvars_1` doesn't work on `let_mono` branch (currently PR #33):

```
let x : ∀ a. (a → a) → (a → a) = let y : a → a = λw.w
                                 in λz.y
in x id_int
```

Failure is caused by incorrectly (?) inferred type of the inner let and its subsequent unification with type signature on `x`. In the working version we have:

```
Generalizable let-binder with type annotation:
  Annotation: {id=8, structure=forall [{id=4, rank=-1}]. {id=7, structure=({id=6, structure=({id=4, rank=-1} -> {id=4, rank=-1}), rank=-1} -> {id=5, structure=({id=4, rank=-1} -> {id=4, rank=-1}), rank=-1}), rank=-1}, rank=1}
  Inferred  : forall [{id=12, rank=-1, mono}, {id=14, rank=-1, mono}]. {id=11, structure=({id=14, rank=-1, mono} -> {id=13, structure=({id=12, rank=-1, mono} -> {id=12, rank=-1, mono}), rank=2}), rank=-1}
```

The inferred type is `forall b c. b -> (c -> c)` and it unifies with signature `forall a. (a -> a) -> (a -> a)`. On the `let_mono` branch we have:

```
Generalizable let-binder with type annotation:
  Annotation: {id=8, structure=forall [{id=4, rank=-1}]. {id=7, structure=({id=6, structure=({id=4, rank=-1} -> {id=4, rank=-1}), rank=-1} -> {id=5, structure=({id=4, rank=-1} -> {id=4, rank=-1}), rank=-1}), rank=-1}, rank=1}
  Inferred  : forall [{id=17, rank=-1, mono}]. {id=11, structure=({id=17, rank=-1, mono} -> {id=13, structure=({id=12, rank=2, mono} -> {id=12, rank=2, mono}), rank=2}), rank=-1}
```

which means we infer type `forall b. b -> (c -> c)` that doesn't unify with the annotation on `x` due to restriction on mixed-prefix unification (we can't unify a quantified variable `a` in the signature with an existential `c` in the inferred type). I don't understand where the difference in inferred types comes from. In the working version when we carry out generalization of the inner let we have:

```
Exiting existential with unification variable {id=13, structure=({id=12, rank=2, mono} -> {id=12, rank=2, mono}), rank=2}
Exiting existential with unification variable {id=14, rank=2, mono}
State after solving, before exiting:
state.young = 2
Pool 1:
id = 1, rank = 1
id = 3, rank = 1
id = 1, rank = 1
id = 1, rank = 1
id = 3, rank = 1
id = 0, rank = 1
Pool 2:
id = 11, rank = 2
id = 13, rank = 2
id = 14, rank = 2
id = 12, rank = 2
id = 13, rank = 2
id = 13, rank = 2
id = 11, rank = 2
Ending generalize.exit
Generalizable vars from the generalization engine: [{id=14, rank=-1, mono}, {id=12, rank=-1, mono}]
Generalizable schemes from the generalization engine: [forall [{id=12, rank=-1, mono}, {id=14, rank=-1, mono}]. {id=11, structure=({id=14, rank=-1, mono} -> {id=13, structure=({id=12, rank=-1, mono} -> {id=12, rank=-1, mono}), rank=2}), rank=-1}]
```
Variables in pool 2 (11,12,13,14) are candidates for being generalized.  Variables 11 and 13 have structure so they can't be quantifiers, variables 12 and 14 don't have structure so they become quantifiers. On the `let_mono` branch we have:

```
Exiting existential with unification variable {id=13, structure=({id=12, rank=2, mono} -> {id=12, rank=2, mono}), rank=2}
Exiting existential with unification variable {id=17, rank=2, mono}
State after solving, before exiting:
state.young = 2
Pool 1:
id = 1, rank = 1
id = 3, rank = 1
id = 1, rank = 1
id = 1, rank = 1
id = 3, rank = 1
id = 0, rank = 1
Pool 2:
id = 17, rank = 2
id = 11, rank = 2
id = 13, rank = 2
id = 17, rank = 2
id = 12, rank = 2
id = 13, rank = 2
id = 13, rank = 2
id = 11, rank = 2
Ending generalize.exit
Generalizable vars from the generalization engine: [{id=17, rank=-1, mono}]
Generalizable schemes from the generalization engine: [forall [{id=17, rank=-1, mono}]. {id=11, structure=({id=17, rank=-1, mono} -> {id=13, structure=({id=12, rank=2, mono} -> {id=12, rank=2, mono}), rank=2}), rank=-1}]
```
Inputs look exactly the same as before, except that variable numbering is different: 11,12,13, and 17. Variables 11 and 13 again form structure of the type, whereas variables 12 and 17 don't have a structure and could be generalized. But this time generalization engine finds only one quantifier: 17. Variable 12 is not considered generalisable and at the same time disappears from the pool. The code responsible for discarding it is in `Generalization.exit`:

```ocaml
let vs =
  List.filter (fun v ->
    U.is_representative v && begin
      if U.rank v < state.young then begin
        register_at_rank state v;
        false
      end
      else begin
        assert (U.rank v = state.young);
        U.set_rank v generic;
        U.structure v = None
      end
    end
  ) vs
```
Variable `12` is filtered out at this stage because condition `U.is_representative` is false for it. I believe it should be the case that for every variable `id` in the pool there is exactly one copy that is representation (i.e. a point of a unionfind algorithm) and all other copies are only links to that point. This doesn't seem to be the case here. In the working version there is also just a single copy of 12, but that one is representative. I have no idea how this situation arises. There is no obvious difference - inferred types of bindings `y` and `z` are identical in both cases.

**Comment by jstolarek** [comment]
Bisect points to 881733da8d49b7c9600e2d85e172ecb9fc501d64

(Note to self: restore comment that got deleted in this commit?)

**Comment by jstolarek** [comment]
TL;DR Fuck...

I got to the bottom of this by printing extra information telling whether a given copy of a variable is representative or not. (A representative variable is the one that describes a point/equivalence class in a unionfind algorithm; all other copies of that variables are links that directly or indirectly point to a representative variable.)

Let's first trace how `y` receives a type in a version thta uses `def` constraints.  Generalization engine infers the following:

```
Pool 3:
id = 17, rank = 3
id = 18, rank = 3, mono
id = 18, rank = 3, mono, repr
id = 17, rank = 3, repr
Ending generalize.exit
Generalizable vars from the generalization engine: [{id=18, rank=-1, mono, repr}]
Generalizable schemes from the generalization engine:
   [forall [{id=18, rank=-1, mono}]. {id=17, structure=({id=18, rank=-1, mono, repr} -> id=18, rank=-1, mono}), rank=-1, repr}]
```

Notice that it picked representative copy of a variable as the generalizable variable and that this representative variable also appears in the type scheme itself - this will become crucial. Then we attempt to unify bodies of the inferred type scheme and the annotation:

```
Unifying let annotation with inferred type of let body. Trying to unifying variables:
  {id=13, structure=({id=12, rank=-1, repr} -> {id=12, rank=-1, repr}), rank=2, repr}
  {id=17, structure=({id=18, rank=-1, mono, repr} -> {id=18, rank=-1, mono}), rank=-1, repr}
Unification successful.  Variable after unification:
  {id=13, structure=({id=12, rank=-1, mono, repr} -> {id=12, rank=-1, mono}), rank=-1}
```

The resulting type again contains a representative copy of variable 12.  Since only representative variables can be generalized everything goes as expected from here. Contrast this with a version that uses let_mono constraints:

```
Pool 3:
id = 18, rank = 3, mono, repr
id = 17, rank = 3
id = 18, rank = 3, mono
id = 18, rank = 3, mono
id = 17, rank = 3, repr
Ending generalize.exit
Generalizable vars from the generalization engine: [{id=18, rank=-1, mono, repr}]
Generalizable schemes from the generalization engine:
  [forall [{id=18, rank=-1, mono}]. {id=17, structure=({id=18, rank=-1, mono} -> {id=18, rank=-1, mono}), rank=-1, repr}]
```

Variables in the pool are ordered slightly differently and as a result type scheme from the generalization does not contain a representative copy of variable 18 (though the generalizable variable is still a representative one). We then unify bodies of the inferred type with a signature:

```
Unifying let annotation with inferred type of let body. Trying to unifying variables:
  {id=13, structure=({id=12, rank=-1, repr} -> {id=12, rank=-1, repr}), rank=2, repr}
  {id=17, structure=({id=18, rank=-1, mono} -> {id=18, rank=-1, mono}), rank=-1, repr}
Unification successful.  Variable after unification:
  {id=13, structure=({id=12, rank=-1, mono} -> {id=12, rank=-1, mono}), rank=-1}
```

Note how during unification we unify a representative copy of variable 12 with nonrepresentative copy of variable 18 and return a non-representative copy of variable 12. This is quite surprising and has a knock-on effect later on: no representative copy of variable 12 makes it into the pool and so it is not considered candidate for generalisation. Now, what dictates this choice of not returning a representative copy? Here's what the comment in TUnionFind says:

> The weights are used to determine whether [point1] should be made to point to [point2], or vice-versa. By making the representative of the smaller class point to that of the larger class, we guarantee that paths remain of logarithmic length (not accounting for path compression, which makes them yet smaller).

If weights are equal choice is made arbitrarily:

```ocaml
if weight1 >= weight2 then begin
  TRef.set t point2 (Link point1);
  TRef.set t point1 (Info (weight1 + weight2, f desc1 desc2))
end
else begin
  TRef.set t point1 (Link point2);
  TRef.set t point2 (Info (weight1 + weight2, f desc1 desc2))
end
```

Interestingly if I change the condition from `>=` to `>` example `scoped_tyvars_1` suddenly works as expected. Of course we can't rely on this - this sort of detail should never leak outside of unionfind algorithm. I placed a fix in the generalisation algorithm (c.f. old version above):

```ocaml
let vs =
  List.filter (fun v ->
    begin
      if U.rank v = generic then
        (* A copy of this variable already visited, discard *)
        false
      else
      if U.rank v < state.young then begin
        register_at_rank state v;
        false
      end
      else begin
        assert (U.rank v = state.young);
        U.set_rank v generic;
        U.structure v = None
      end
    end
  ) vs
```
This possibly makes paths traversed by unionfind algorithm a bit longer but at least it's correct.
